azurelinuxagent/ga/agent_update_handler.py

```diff
@@ -160,38 +198,142 @@ def __get_agent_family_manifests(self, goal_state):
+            logger.warn(msg)
+            add_event(op=WALAEventOperation.AgentUpgrade, is_success=False, message=msg, log_event=True)
+        return None
 
-    def __download_and_get_agent(self, goal_state, agent_family, agent_manifest, requested_version):
+
+class GAVersionUpdater(object):
+
+    def __init__(self, gs_id, agent_family, agent_manifest, version, update_state):
+        self._gs_id = gs_id
+        self._agent_family = agent_family
+        self._agent_manifest = agent_manifest
+        self._version = version
+        self._update_state = update_state
+
+    def should_update_agent(self, goal_state):
         """
-        This function downloads the new agent(requested version) and returns the downloaded version.
+        RSM version update:
+            update is allowed once per (as specified in the conf.get_autoupdate_frequency()) and
+            if new version not same as current version, not below than daemon version and if version is from rsm request
+            return false when we don't allow updates.
+        self-update:
+            1) checks if we allowed download manifest as per manifest download frequency
+            2) update is allowed once per (as specified in the conf.get_self_update_hotfix_frequency() or conf.get_self_update_regular_frequency())
+            3) not below than current version
+            return false when we don't allow updates.
+        """
+        raise NotImplementedError
+
+    def log_new_agent_update_message(self):
+        """
+        This function logs the update message after we check agent allowed to update.
+        """
+        raise NotImplementedError
+
+    def purge_extra_agents_from_disk(self):
+        """
+        RSM version update:
+            remove the agents( including rsm version if exists) from disk except current version. There is a chance that rsm version could exist and/or blacklisted
+            on previous update attempts. So we should remove it from disk in order to honor current rsm version update.
+        self-update:
+            remove the agents from disk except current version and new agent version if exists
+        """
+        raise NotImplementedError
+
+    def proceed_with_update(self):
         """
-        if agent_manifest is None:  # Fetch agent manifest if it's not already done
-            agent_manifest = goal_state.fetch_agent_manifest(agent_family.name, agent_family.uris)
-        package_to_download = self.__get_agent_package_to_download(agent_manifest, requested_version)
+        RSM version update:
+                upgrade/downgrade to the specified version.
+                Raises: AgentUpgradeExitException
+        self-update:
+                If largest version is found in manifest, upgrade to that version. Downgrade is not supported.
+                Raises: AgentUpgradeExitException
+        """
+        raise NotImplementedError
+
+    def download_and_get_new_agent(self, protocol, goal_state):
```

Comment:
Another issue with the design of this method is that it takes the goal state as parameter and the init method takes several properties from the goal state. With this design, this goal state and the properties in init may correspond to different goal states. You may want to consider passing the goal state on init and remove it from here